### PR TITLE
PoC of "additional layer store" of Podman/CRI-O and lazy pulling of zstd:chunked

### PR DESCRIPTION
**Pull request by ktock**
This PR contains the following:

- PoC of "additional layer store" of Podman/CRI-O
- PoC of lazy pulling of zstd:chunked

[Zstd:chunked](https://github.com/containers/storage/pull/775) is a variant of zstd that enables extracting each file separately. The structure based on stargz and its verification logic is based on  eStargz. Currently, this is enabled by directly patching `estargz` lib for supporting that compression algorithm (zstd) and that footer format. Maybe we should make `estargz` lib configurable against compression algorithms and separate zstd:chunked patches into an indipendent package.

["Additional layer store"](https://github.com/containers/storage/pull/795) is a proposed functionality for https://github.com/containers/storage to provide remotely mounted layers to the runtime. Currently, this is implemented as a thin wrapper of our [`fs`](https://github.com/containerd/stargz-snapshotter/tree/master/fs) package. Instead of implementing snapshotter API, this directly exposes layers on the filesystem. This seems a bit out of scope as a non-core subproject of containerd. So this should be separated to another project, maybe?

The following command mounts the store at the `<mountpoint>`:

```
# registry-storage <mountpoint>
```

This store exposes the following structure required as an additional layer store:

```
<mountpoint>/base64(imageref)/<layerdigest>/
- layer : exposes the extracted layer
- info  : contains JSON-formatted "Layer" structure of containers/storage
- use   : files to notify the use of this layer (used for management of reference count)
```

There are also other APIs for debugging:

```
<mountpoint>/base64(imageref)/
- manifest       : contains manifest of this image
- config         : contains config blob of this image
- <layerdigest>/
  - chain        : exposes the extracted chain from the bottommost to this layer
```

### HelloBench result of Podman + {eStargz, zstd:chunked} on GitHub Actions (`Standard_DS2_v2` instance at `eastus2`):

https://github.com/ktock/stargz-snapshotter/actions/runs/623653929

I'll work on experiment on a richer instance with larger number of samples.

result

### CODE REVIEW

**Comment by ktock** [comment]
cc @giuseppe @AkihiroSuda

File: cmd/ctr-remote/commands/convert.go [hunk 1]
```diff
@@ -66,6 +67,11 @@ When '--all-platforms' is given all images in a manifest list must be available.
 			Usage: "eStargz chunk size",
 			Value: 0,
 		},
+		// zstd:chunked flags
+		cli.BoolFlag{
+			Name:  "zstdchunked",
```
Comment (AkihiroSuda) [inline cmd/ctr-remote/commands/convert.go:72]: nit: `zstdchunked` -> `zstd-chunked` (just my preference)

File: cmd/registry-storage/main.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	"github.com/containerd/stargz-snapshotter/fs/source"
+	"github.com/containerd/stargz-snapshotter/service/keychain"
+	fusefs "github.com/hanwen/go-fuse/v2/fs"
+	"github.com/hanwen/go-fuse/v2/fuse"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	defaultLogLevel = logrus.InfoLevel
+	defaultRootDir  = "/var/lib/registry-storage"
```
Comment (AkihiroSuda) [inline cmd/registry-storage/main.go:44]: Shouldn't this be something like `/var/lib/stargz-registry-storage`?

**Comment by AkihiroSuda** [comment]
> # registry-storage <mountpoint>

Shouldn't the exec name be like `stargz-registry-storage`

**Comment by ktock** [comment]
Closing in favor of https://github.com/containerd/stargz-snapshotter/pull/293 and https://github.com/containerd/stargz-snapshotter/pull/301.

**Pull request closed without merging**